pkg/engine/events.go

```diff
@@ -188,13 +188,62 @@ func makeEventEmitter(events chan<- Event, update UpdateInfo) (eventEmitter, err
 
 	logging.AddGlobalFilter(logging.CreateFilter(secrets, "[secret]"))
 
+	// Instead of sending to the source channel directly, buffer events to account for slow receivers.
+	//
+	// Buffering is done by a goroutine that concurrently receives from the senders and attempts to send events to the
+	// receiver. Events that are received while waiting for the receiver to catch up are buffered in a slice.
+	//
+	// We do not use a buffered channel because it is empirically less likely that the goroutine reading from a
+	// buffered channel will be scheduled when new data is placed in the channel.
+	buffer, done := make(chan Event), make(chan bool)
+	go func() {
+		defer close(done)
+
+		var queue []Event
+		for {
+			contract.Assert(buffer != nil)
+
+			e, ok := <-buffer
+			if !ok {
+				return
+			}
+			queue = append(queue, e)
+
+			// While there are events in the queue, attempt to send them to the waiting receiver. If the receiver is
+			// blocked and an event is received from the event senders, stick that event in the queue.
+			for len(queue) > 0 {
+				select {
+				case e, ok := <-buffer:
+					if !ok {
+						// If the event source has been closed, flush the queue.
+						for _, e := range queue {
+							events <- e
+							if e.Type == CancelEvent {
+								return
+							}
+						}
+						return
+					}
+					queue = append(queue, e)
```

Comment:
I suppose this can theoretically grow unboundedly?  Is there any real risk there?  I'm not sure quite what it would take to cause that to happen.

Reply:
I don't think that there's a realistic risk there. It would take total starvation of the send half of this goroutine for that to happen.